.github/workflows/coverage.yml

```diff
@@ -0,0 +1,28 @@
+name: Coverage
+
+on: [push, pull_request]
+jobs:
+  coverage:
+    runs-on: ubuntu-20.04
+    steps:
+      - uses: actions/checkout@v2
+      - name: dependencies
+        shell: bash
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y -qq make cmake gcc g++ lcov
+      - name: build
+        shell: bash
+        run: |
+          export LD_LIBRARY_PATH=.:$(cat /etc/ld.so.conf.d/* | grep -vF "#" | tr "\\n" ":" | sed -e "s/:$//g")
+          cmake . -DCMAKE_BUILD_TYPE=Debug -DENABLE_COVERAGE=On
+          make all -j$(nproc)
+      - name: coverage
+        shell: bash
+        run: |
+          make coverage
+          lines=$(lcov --summary coverage.info | grep -F lines | awk '{print $2}' | sed -e "s/%//g")
+          if (( $(echo "${lines} < 85.0" | bc -l) )); then
```

Comment:
we require that the project stays above 85% of all lines covered by the test. let me know if this sounds ok to you or not. currently we are at 89.1% and i think with a little more tedious testing we could get to 95%.

Reply:
85% is fine. We should probably have it standing out a bit more using an environment variable:
```
      - name: coverage
        shell: bash
        run: |
          make coverage
          lines=$(lcov --summary coverage.info | grep -F lines | awk '{print $2}' | sed -e "s/%//g")
          if (( $(echo "${lines} < $COVERAGE_THRESHOLD" | bc -l) )); then
              echo "Line coverage dropped below $COVERAGE_THRESHOLD% to ${lines}%"
              exit 1
          fi
        env:
          COVERAGE_THRESHOLD: 85.0
```

Reply:
Let's also add a target in the Makefile:
```
BUILD_MODE ?= Release
ENABLE_COVERAGE ?= On

build: clean
	@echo "Build in ${BUILD_MODE} mode"
	mkdir -p bin/${BUILD_MODE}
	@cd bin/${BUILD_MODE}; cmake ../../ -DCMAKE_BUILD_TYPE=${BUILD_MODE} -DENABLE_COVERAGE=${ENABLE_COVERAGE}
	@cd bin/${BUILD_MODE}; make

test: build
	@cd bin/${BUILD_MODE}; make test

coverage: build
	@cd bin/${BUILD_MODE}; make coverage

install: build
	@cd bin/${BUILD_MODE}; make install

clean:
	@rm -rf bin
```